scrapy/dispatch/dispatcher.py

```diff
@@ -0,0 +1,386 @@
+                    response = receiver(signal=self, sender=sender, **named)
+                else:
+                    response = robust_apply(receiver, sender=sender, **named)
+                if isinstance(response, Deferred):
+                    logger.error("Cannot return deferreds from signal handler: %(receiver)s",
+                                 {'receiver': receiver}, extra={'spider': spider})
+            except dont_log:
+                response = Failure()
+            except Exception:
+                response = Failure()
+                logger.error("Error caught on signal handler: %(receiver)s",
+                             {'receiver': receiver},
+                             exc_info=True, extra={'spider': spider})
+            responses.append((receiver, response))
+        return responses
+
+    def send_robust_deferred(self, sender, **named):
+        """
+        Send signal from sender to all connected receivers catching errors.
+        Modified to work with functionons returning twisted deferreds.
+        Arguments:
+
+            sender
+                The sender of the signal. Can be anything python object
+                (normally one registered with a connect if you actually
+                 want something to occur).
+
+            named
+                Named arguments which will be passed to receivers. These
+                arguments must be a subset of the argument names defined in
+                providing_args.
+        """
+        dont_log = named.pop('dont_log', _IgnoredException)
+
+        def logerror(failure, recv):
+            spider = named.get('spider', None)
+            if dont_log is None or not isinstance(failure.value, dont_log):
+                logger.error("Error caught on signal handler: %(receiver)s",
+                             {'receiver': recv},
+                             exc_info=failure_to_exc_info(failure),
+                             extra={'spider': spider})
+            return failure
+        dfds = []
+        if not self.receivers or self.sender_receivers_cache.get(sender) is NO_RECEIVERS:
+            return dfds
+        # Call each receiver with whatever arguments it can accept.
+        # Return a list of tuple pairs [(receiver, response), ... ].
+        for receiver in self._live_receivers(sender):
+            dfd = maybeDeferred(
+                receiver,  signal=self, sender=sender, **named)
+            dfd.addErrback(logerror, receiver)
+            dfd.addBoth(lambda result: (receiver, result))
+            dfds.append(dfd)
+        d = DeferredList(dfds)
+        d.addCallback(lambda out: [x[1] for x in out])
+        return d
+
+    def _clear_dead_receivers(self):
+        # Note: caller is assumed to hold self.lock.
+        if self._dead_receivers:
+            self._dead_receivers = False
+            new_receivers = []
+            for r in self.receivers:
+                if isinstance(r[1], weakref.ReferenceType) and r[1]() is not None:
```

Comment:
This `for` loop used to contain:
```python
                if isinstance(r[1], weakref.ReferenceType) and r[1]() is None:
                    continue
                new_receivers.append(r)
```
(i.e. everything that was not a dead weakref was added to `new_receivers`) With the changes you made it looks as if `new_receivers` will almost always be empty, except for live weakrefs. You probably meant to do this, right?
```python
                if not (isinstance(r[1], weakref.ReferenceType) and r[1]() is None):
```